File: backend/app/schemas/master.py
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field

class DataFrequency(str, Enum):
    SHIFT = "SHIFT"
    DAILY = "DAILY"
    WEEKLY = "WEEKLY"
    MONTHLY = "MONTHLY"

class StandardIdentifiers(BaseModel):
    date: Optional[str] = Field(None, description="ISO date YYYY-MM-DD")
    timestamp: Optional[str] = Field(None, description="ISO timestamp")
    shift: Optional[str] = Field(None, description="Shift identifier e.g., Shift A, Shift B, Shift 1")
    unit_id: Optional[str] = Field(None, description="Unit I, Unit II")
    department_id: Optional[str] = Field(None, description="Spinning, Sizing, Weaving, etc.")
    section_id: Optional[str] = Field(None, description="Section A, Section B")
    process_id: Optional[str] = Field(None, description="Carding, Draw Frame, Simplex, Weaving, Sizing")
    machine_type_id: Optional[str] = Field(None, description="Ring Frame, Vortex, Airjet, Toyota")
    machine_id: Optional[str] = Field(None, description="SMX-03, V-05, etc.")
    product_id: Optional[str] = None
    batch_id: Optional[str] = None
    order_id: Optional[str] = None
    employee_id: Optional[str] = None

class RawVsDerivedSchema(BaseModel):
    raw_target: float = Field(..., description="Target production from report")
    raw_actual: float = Field(..., description="Actual production recorded")
    
    @property
    def derived_production_loss(self) -> float:
        return max(0.0, self.raw_target - self.raw_actual)

    @property
    def derived_achievement_percentage(self) -> float:
        if self.raw_target <= 0:
            return 0.0
        return round((self.raw_actual / self.raw_target) * 100.0, 2)

File: frontend/src/types/master.ts
export type DataFrequency = 'SHIFT' | 'DAILY' | 'WEEKLY' | 'MONTHLY';

export interface StandardIdentifiers {
  date?: string;
  timestamp?: string;
  shift?: string;
  unit_id?: string;
  department_id?: string;
  section_id?: string;
  process_id?: string;
  machine_type_id?: string;
  machine_id?: string;
  product_id?: string;
  batch_id?: string;
  order_id?: string;
  employee_id?: string;
}

export interface RawVsDerived {
  raw_target: number;
  raw_actual: number;
  derived_production_loss: number;
  derived_achievement_percentage: number;
}
